File: templates/javaee-app/integration-tests/src/test/java/org/sgodden/tom/integration/services/CustomerServiceTest.java
package org.sgodden.tom.integration.services;

import org.sgodden.tom.integration.AbstractIntegrationTest;
import org.sgodden.tom.model.CustomerOrderStatus;
import org.sgodden.tom.model.ICustomerOrder;
import org.sgodden.tom.services.customerorder.CustomerOrderListEntry;
import org.sgodden.tom.services.customerorder.CustomerOrderService;
import org.springframework.beans.factory.annotation.Autowired;
import org.testng.annotations.AfterMethod;
import org.testng.annotations.Test;

import static org.testng.Assert.assertEquals;

@Test
public class CustomerServiceTest extends AbstractIntegrationTest {

    @Autowired
    private CustomerOrderService customerOrderService;

    @AfterMethod
    public void afterTest() {
        for (CustomerOrderListEntry order : customerOrderService.list()) {
            customerOrderService.remove(order.getId());
        }
    }

    public void testCreate() {
        ICustomerOrder order = customerOrderService.create();
        assertEquals(order.getStatus(), CustomerOrderStatus.NEW);
    }

    public void testPersist() {
        createOrder(1);
        assertEquals(customerOrderService.list().size(), 1);
    }

    public void testFindAll() {
        for (int i = 0; i < 10; i++) {
            createOrder(i + 1);
        }
        assertEquals(customerOrderService.list().size(), 10);
    }

    private void createOrder(int seq) {
        ICustomerOrder order = customerOrderService.create();
        order.setCustomerReference("REFERENCE: " + seq);
        order.setOrderNumber("ORDER NUMBER: " + seq);
        customerOrderService.persist(order);
    }

}

File: templates/scala-mongodb/integration-tests/src/test/scala/org/sgodden/tom/integration/services/CustomerServiceTest.scala
package org.sgodden.tom.integration.services

import org.sgodden.tom.services.customerorder.CustomerOrderService
import org.testng.Assert._
import org.sgodden.tom.model.{ValidationException, CustomerOrderStatus, ICustomerOrder}
import org.springframework.beans.factory.annotation.Autowired
import org.sgodden.tom.integration.AbstractIntegrationTest
import org.testng.annotations.{BeforeMethod, Test}

@Test
class CustomerServiceTest extends AbstractIntegrationTest {

  @Autowired private var customerOrderService: CustomerOrderService = null

  @BeforeMethod def beforeTest {
    for (order <- customerOrderService.findAll) {
      customerOrderService.remove(order.getId)
    }
  }

  def testCreate {
    var order: ICustomerOrder = customerOrderService.create
    assertEquals(order.getStatus, CustomerOrderStatus.NEW)
  }

  def testPersist {
    createOrder(1)
    assertEquals(customerOrderService.findAll.size, 1)
  }

  def testInvalidOrderIsNotPersisted {
    val order = makeOrder(1);
    order.setCustomerReference("wrong")
    try {
      customerOrderService.persist(order)
    } catch {
      case ve: ValidationException => {}
    }
    assertEquals(customerOrderService.findAll.size, 0)
  }

  def testFindAll {
    (1 to 10).foreach(createOrder(_))
    assertEquals(customerOrderService.findAll.size, 10)
  }

  @Test
  def shouldFailWithNullOrderNumber {
    var order: ICustomerOrder = customerOrderService.create
    try {
      customerOrderService.persist(order)
      fail("Validation exception should have been thrown")
    } catch {
      case ex: ValidationException => {}
    }
  }

  private def createOrder(seq: Int): Unit = {
    customerOrderService.persist(makeOrder(seq))
  }

  private def makeOrder(seq: Int) = {
    var order: ICustomerOrder = customerOrderService.create
    order.setCustomerReference("crREFERENCE: " + seq)
    order.setOrderNumber("ORDER NUMBER: " + seq)
    order
  }
}
